### ISSUE TITLE
IF6080 Pembelajaran Mesin

### ISSUE THREAD

**Issue by banditelol** [banditelol/SistemIntelijenS2ITB#3]
# Pembelajaran Mesin

**Comment by banditelol** [comment]
## Andrew Ng's Structure ML Project

### Avoidable Bias
![image](https://user-images.githubusercontent.com/5263688/99925043-1ad94b00-2d6f-11eb-8a69-da4453a38da2.png)

Training error vs Human Error has big gap, it's indicative that training is not fitted well enough so focus on bias reduction
But when training error - human is inisignificant, we need to focus on variance reduction between training and dev (regularization, etc)
Human Level error as proxy for bayes error.

Human - Training : Avoidable Bias
Training - Dev : Variance

### Understanding Human Level Performance

Human-level error is a proxy for Bayes Error. Medical Image classification example

1. Typical Human : 3%
2. Typical Doctor : 1% 
3. Experienced Doctor 0.7 %
4. Team of Experienced Doctors : 0.5%

What is human level error in this case?
To have a proxy for bayes error, so that we use the smallest prior error. In this case Team of experienced doctors.

For Publishing and Production, it may have different context and purpose. And we could argue that typical doctor is a good estimate for human-level error.
![image](https://user-images.githubusercontent.com/5263688/99925384-8a9c0580-2d70-11eb-8716-5ca539b05272.png)

It is important to estimate bayes error to understand how well we really do in a problem

### Surpassing Human-Level Performance
![image](https://user-images.githubusercontent.com/5263688/99925552-2a599380-2d71-11eb-84f4-692f4bfaa6d6.png)

once we surpassed the human-level performance, it's harder to make progress because we don't know if its really good or overfitting, and we don't know how much avoidable bias vs variance to focus on

Several model which surpassed human mostly are not perception task.
![image](https://user-images.githubusercontent.com/5263688/99925581-4bba7f80-2d71-11eb-8007-d2f1373ca638.png)

several properties:
- Structured Data
- Not perception
- Lots of Data

> Question
> When we're talking surpassing human level performance, what measure of performace should we care about?

### Improve Model Performance
![image](https://user-images.githubusercontent.com/5263688/99925701-aeac1680-2d71-11eb-82e6-b8e51bc1633c.png)

Reducing avoidable bias and variance. First reduce avoidable bias, then focus on reducing variance. There's several tactics here

#### Avoidable Bias

- Train Bigger Model
- Train longer
- better optimization algo
  - momentum, RMSprop, Adam
- NN architecture/ hyperparameter search

#### Variance

- More data to generalize better
- Regularization
  - L2
  - Dropout
  - data augmentation
- NN architecture and hyperparameter search

> Question
> So NN arch and hyperparameter affects multiple measures

### Peacetopia Quiz

Question 1
Problem Statement
This example is adapted from a real production application, but with details disguised to protect confidentiality.

You are a famous researcher in the City of Peacetopia. The people of Peacetopia have a common characteristic: they are afraid of birds. To save them, you have to build an algorithm that will detect any bird flying over Peacetopia and alert the population.

The City Council gives you a dataset of 10,000,000 images of the sky above Peacetopia, taken from the city’s security cameras. They are labelled:

- y = 0: There is no bird on the image
- y = 1: There is a bird on the image

Your goal is to build an algorithm able to classify new images taken by security cameras from Peacetopia.

There are a lot of decisions to make:

- What is the evaluation metric?
- How do you structure your data into train/dev/test sets?

#### Metric of Success

The City Council tells you that they want an algorithm that

1. Has high accuracy
2. Runs quickly and takes only a short time to classify a new image.
3. Can fit in a small amount of memory, so that it can run in a small processor that the city will attach to many different security cameras.

Note: Having three evaluation metrics makes it harder for you to quickly choose between two different algorithms, and will slow down the speed with which your team can iterate. True/False?

**TRUE**

After further discussions, the city narrows down its criteria to:

- "We need an algorithm that can let us know a bird is flying over Peacetopia as accurately as possible."
- "We want the trained model to take no more than 10sec to classify a new image.”
- “We want the model to fit in 10MB of memory.”

If you had the three following models, which one would you choose?
![image](https://user-images.githubusercontent.com/5263688/99926007-bcae6700-2d72-11eb-8385-108b7b7b8053.png)

because two of them are satisficing metrics, so it needs only to be fulfilled, not optimized, while test accuracy need to be optimized as long as the other metrics are satisfied.

![image](https://user-images.githubusercontent.com/5263688/99926056-e8c9e800-2d72-11eb-8931-297727ce1d9d.png)

#### Structuring Your Data

Before implementing your algorithm, you need to split your data into train/dev/test sets. Which of these do you think is the best choice?

![image](https://user-images.githubusercontent.com/5263688/99926078-f97a5e00-2d72-11eb-87e5-50ab879ca4be.png)

because of the sheer amount of data, it needs not to be the conventional 70:30 or 60:20:20 split, it's safe to say 95:2.5:2.5 is still a safe bet. Maybe we could refer to the whole talks about VC dimension to get a more quantitative views.

Question 5
After setting up your train/dev/test sets, the City Council comes across another 1,000,000 images, called the “citizens’ data”. Apparently the citizens of Peacetopia are so scared of birds that they volunteered to take pictures of the sky and label them, thus contributing these additional 1,000,000 images. These images are different from the distribution of images the City Council had originally given you, but you think it could help your algorithm.

Notice that adding this additional data to the training set will make the distribution of the training set different from the distributions of the dev and test sets.  

Is the following statement true or false?

"You should not add the citizens' data to the training set, because if the training distribution is different from the dev and test sets, then this will not allow the model to perform well on the test set."

**TRUE** if we wanted to add the data, make sure it is stratisfied, so that the distribution of train, dev, test comes from the same distribution

One member of the City Council knows a little about machine learning, and thinks you should add the 1,000,000 citizens’ data images to the test set. You object because:

![image](https://user-images.githubusercontent.com/5263688/99926419-3561f300-2d74-11eb-9163-74e4dc00680c.png)

You train a system, and its errors are as follows (error = 100%-Accuracy):

Error | Value
-- | --
Training set error | 4.0%
Dev set error | 4.5%

This suggests that one good avenue for improving performance is to train a bigger network so as to drive down the 4.0% training error. Do you agree?

![image](https://user-images.githubusercontent.com/5263688/99926472-71955380-2d74-11eb-827b-31436179036b.png)

because to have justification on reducing the bias, we need to know the bayes error or at least its approximately

![image](https://user-images.githubusercontent.com/5263688/99926509-94c00300-2d74-11eb-9813-608f15093232.png)

because its assumed it matter of great importance, so that it is also important to have the best measure as its reference of human-level performance.

![image](https://user-images.githubusercontent.com/5263688/99926563-bae5a300-2d74-11eb-9398-2cf7aba95975.png)

> Question
> What's the definition of bayes error anyway? I just let it flow and assume its absolute error

![image](https://user-images.githubusercontent.com/5263688/99926619-e799ba80-2d74-11eb-8a8d-32d58220de49.png)

We need to decrease bias error, so regularization is not the way to go (it decrease variance) and bigger training set to reduce variance is not our goal here (as the goal is reducing bias). Thus the last one is the only sensible option.

![image](https://user-images.githubusercontent.com/5263688/99926750-50813280-2d75-11eb-920e-afa0270ac003.png)

because we could see that training set error is actually underfitted compared to human level performance, and test set performance is significantly lower, it's indicating that dev set doesn't represent the test set, so we need to try to get better dev set which comes from the same distribution as test set.

![image](https://user-images.githubusercontent.com/5263688/99926833-9dfd9f80-2d75-11eb-815f-b2ca95e1213f.png)

As the avoidable bias is not clear (train error is lower than human error) its hard to get better intuition on how to do better. And to make sure that its not over fitting, if the test set is big enough, its implying that our bayes error is lower than 0.05%

![image](https://user-images.githubusercontent.com/5263688/99926918-ee74fd00-2d75-11eb-894b-47ab46ac307a.png)

its self explanatory.

Question 14
You’ve handily beaten your competitor, and your system is now deployed in Peacetopia and is protecting the citizens from birds! But over the last few months, a new species of bird has been slowly migrating into the area, so the performance of your system slowly degrades because your data is being tested on a new type of data.

You have only 1,000 images of the new species of bird. The city expects a better system from you within the next 3 months. Which of these should you do first?

![image](https://user-images.githubusercontent.com/5263688/99926965-1c5a4180-2d76-11eb-8da9-f03503c604d8.png)

I don't know the intution on what's best to fo **first**. But certainly to balance the dataset we need to augment the data. But if the existing data actually has around 1000 images per species then it's a good idea to add that image into the dataset and split it again.

Question 15
The City Council thinks that having more Cats in the city would help scare off birds. They are so happy with your work on the Bird detector that they also hire you to build a Cat detector. (Wow Cat detectors are just incredibly useful aren’t they.) Because of years of working on Cat detectors, you have such a huge dataset of 100,000,000 cat images that training on this data takes about two weeks. Which of the statements do you agree with? (Check all that agree.)

![image](https://user-images.githubusercontent.com/5263688/99927023-604d4680-2d76-11eb-9432-262caf9669c9.png)

**Comment by banditelol** [comment]
## Andrew Ng Improving DNN

### Normalizing inputs

![image](https://user-images.githubusercontent.com/5263688/99951058-0cf7ea00-2db0-11eb-9b05-98496713ef90.png)

subtract out mean and normalize variance. But remember to also normalize the dev set with training data mean and variance

> Question
> How about the test set? should we also normalize it?

#### Why normalize?

the contour can be very elongated and harder to do gradient descent compared to spherical and easier to optimize.
![image](https://user-images.githubusercontent.com/5263688/99951215-4c263b00-2db0-11eb-8085-5d2d2afeba60.png)
